cts/cts-regression.in

```diff
@@ -1,230 +1,335 @@
+
+    :param components: A list of names of components for which to run
+        tests.
+    :type tests: list[str]
+    :return: :attr:`CrmExit.OK` if all tests were successful,
+        :attr:`CrmExit.ERROR` otherwise.
+    :rtype: :class:`CrmExit`
+    """
+    failed = []
+
+    for comp in components:
+        rc = comp.run(verbose, valgrind)
+        if rc != CrmExit.OK:
+            error_print('%s regression tests failed (%s)' % (comp.name, rc))
+            failed.append(comp.name)
+
+    if failed:
+        print('Failed regression tests:', end='')
+        for comp in failed:
+            print(' %s' % comp, end='')
+        print()
+        return CrmExit.ERROR
+
+    return CrmExit.OK
+
+
+def main():
+    """Run Pacemaker regression tests as specified by arguments."""
+    global TEST_HOME
+    try:
+        TEST_HOME = os.path.dirname(os.readlink(sys.argv[0]))
+    except OSError:
+        TEST_HOME = os.path.dirname(sys.argv[0])
+
+    # Available components
+    components = {
+        'cli': Component(
+            'cli',
+            'Command-line tools',
+            use_sudo=False,
+            supports_valgrind=True,
+        ),
+        'exec': Component(
+            'exec',
+            'Local resource agent executor',
+            use_sudo=True,
+            supports_valgrind=False,
+        ),
+        'fencing': Component(
+            'fencing',
+            'Fencer',
+            use_sudo=True,
+            supports_valgrind=False,
+        ),
+        'scheduler': Component(
+            'scheduler',
+            'Action scheduler',
+            use_sudo=False,
+            supports_valgrind=True,
+        ),
+    }
+
+    if REMOTE_ENABLED:
+        components['pacemaker_remote'] = Component(
+            'pacemaker_remote',
+            'Resource agent executor in remote mode',
+            use_sudo=True,
+            supports_valgrind=False,
+        )
+
+    # Build up program description
+    description = textwrap.dedent('''\
+        Run Pacemaker regression tests.
+
+        Available components (default components are 'scheduler cli'):
+    ''')
+
+    for name, comp in sorted(components.items()):
+        description += '\n {:<20} {}'.format(name, comp.description)
+
+    description += (
+        '\n {:<20} Synonym for "scheduler cli exec fencing"'.format('all')
+    )
+
+    # Parse the arguments
+    parser = argparse.ArgumentParser(
+        description=description,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+
+    choices = sorted(components.keys()) + ['all', 'default']
+
+    parser.add_argument('-V', '--verbose', action='store_true',
+                        help='Increase test verbosity')
+    parser.add_argument('-v', '--valgrind', action='store_true',
+                        help='Run test commands under valgrind')
+    parser.add_argument('components', nargs='*', choices=choices,
+                        default='default',
```

Comment:
By the way: I went with `'default'` because we need SOME kind of explicit default value if we use `choices`. We don't have to advertise the literal `'default'` in our description. But for an invalid argument, `argparse` will print the default value as one of the choices. Other options are `''` and `[]`, both of which feel misleading.
```
# with default='default' and 'default' in choices
$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', 'default')
$ cts/cts-regression default
['cli', 'scheduler']

# with default=[] and [] in choices
# choices is no longer all strings, and error message incorrectly suggests user can pass [] from CLI
$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', [])
$ cts/cts-regression []
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: '[]' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', [])

# with default='' and '' in choices
# error message incorrectly suggests user can pass '' from CLI
[reid@laptop pacemaker]$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', '')
[reid@laptop pacemaker]$ cts/cts-regression ''
Traceback (most recent call last):

# All of the above work fine for "no argument on CLI"
```

It might be possible to get rid of `choices` altogether and handle everything in the custom action, but that's less obvious and would require hard-coding the choices into action class (can't pass constructor args manually AFAICT).

Reply:
'default' works for me, otherwise we could do something like 'essential', 'basic', or 'nonroot'